src/control/lib/netdetect/netdetect.go

```diff
@@ -73,91 +76,102 @@ func initLib(xmlTopology string) (C.hwloc_topology_t, error) {
-// where dev is the name of the network device and cpuset and nodeset are
-// the corresponding bitmap strings.
-//
 // Network device names that are not found in the topology are ignored.
 // The order of network devices in the return string depends on the natural
 // order in the system topology and does not depend on the order specified
 // by the input string.
-func GetAffinityForNetworkDevices(netNames string,
-	xmlTopology string) (string, error) {
-
-	var affinity string
+func GetAffinityForNetworkDevices(deviceNames []string) (*[]DeviceAffinity, error) {
+	var affinity []DeviceAffinity
 	var node C.hwloc_obj_t
 	var i C.uint
 	var cpuset *C.char
 	var nodeset *C.char
 
-	topology, err := initLib(xmlTopology)
+	topology, err := initLib()
 	if (err != nil) {
 		log.Debugf("Error from initLib %v", err)
-		return affinity,
+		return nil,
 			errors.New("unable to initialize hwloc library")
 	}
 
 	depth := C.hwloc_get_type_depth(topology, C.HWLOC_OBJ_OS_DEVICE)
 	if depth != C.HWLOC_TYPE_DEPTH_OS_DEVICE {
-		return affinity,
+		return nil,
 			errors.New("hwloc_get_type_depth returned invalid value")
 	}
 
+	netNames := make(map[string]struct{})
+	for _, deviceName := range deviceNames {
+		netNames[deviceName] = struct{}{}
+	}
+
 	numObj := C.hwloc_get_nbobjs_by_depth(topology, C.uint(depth))
 	// for any OS object found in the network device list,
 	// detect and store the cpuset and nodeset of the ancestor node
 	// containing this object
 	for i = 0; i < numObj; i++ {
 		node = C.hwloc_get_obj_by_depth(topology, C.uint(depth), i)
-		if node != nil && strings.Contains(netNames, C.GoString(node.name)) {
-			ancestorNode := C.hwloc_get_non_io_ancestor_obj(topology, node)
-			if ancestorNode == nil {
-				continue
-			}
-			cpusetLen := C.hwloc_bitmap_asprintf(&cpuset,
-				ancestorNode.cpuset)
-			nodesetLen := C.hwloc_bitmap_asprintf(&nodeset,
-				ancestorNode.nodeset)
-			if cpusetLen > 0 && nodesetLen > 0 {
-				affinity += C.GoString(node.name) + ":" +
-					C.GoString(cpuset) + ":" + C.GoString(nodeset) + ";"
-			}
-			C.free(unsafe.Pointer(cpuset))
-			C.free(unsafe.Pointer(nodeset))
+		if node == nil {
+			continue
+		}
+		if _, found := netNames[C.GoString(node.name)]; !found {
+			continue
 		}
+		ancestorNode := C.hwloc_get_non_io_ancestor_obj(topology, node)
+		if ancestorNode == nil {
+			continue
+		}
+		cpusetLen := C.hwloc_bitmap_asprintf(&cpuset,
+			ancestorNode.cpuset)
+		nodesetLen := C.hwloc_bitmap_asprintf(&nodeset,
+			ancestorNode.nodeset)
+		if cpusetLen > 0 && nodesetLen > 0 {
+			var deviceNode DeviceAffinity
+			deviceNode.DeviceName = C.GoString(node.name)
+			deviceNode.CPUSet = C.GoString(cpuset)
+			deviceNode.NodeSet = C.GoString(nodeset)
+			affinity = append(affinity, deviceNode)
+		}
+		C.free(unsafe.Pointer(cpuset))
+		C.free(unsafe.Pointer(nodeset))
 	}
 	cleanUp(topology)
-	return strings.TrimSuffix(affinity, ";"), nil
+	return &affinity, nil
```

Comment:
Echo to the comment on the function signature: You almost never want to do anything with pointers to slices. This shouldn't be returning one.

Reply:
ok, I will fix with the related comment.